kyaml/yaml/walk/associative_sequence.go

```diff
@@ -125,26 +163,36 @@ func (l *Walker) walkAssociativeSequence() (*yaml.RNode, error) {
 		return nil, err
 	}
 
-	// get the merge key from schema
-	var key, strategy string
+	// get the merge key(s) from schema
+	var strategy string
+	var keys []string
 	if l.Schema != nil {
-		strategy, key = l.Schema.PatchStrategyAndKey()
+		strategy, keys = l.Schema.PatchStrategyAndKeyList()
```

Comment:
At this point, if openAPI returns multiple keys, why can't we just decide upon the merge key here itself so that the remaining logic can remain unchanged. 

In simple words, we used to get 1(or 0) merge key from openAPI schema. Now we might get multiple. If we get multiple of those, since we have access to l.Sources here, why can't we just pick the merge key here itself?

Reply:
If I'm understanding your question correctly, the goal is not to choose a single merge key; it's to use all the merge keys from the schema as a tuple - e.g. for containerPort ports it would be the {containerPort, protocol} tuple, and the test cases I added as part of its PR document the desired behavior

Reply:
right, the primary key for a merge can now be a list of fields instead of just one.

we should aspire to treat the old behavior as the degenerate case of a list of length one.

Reply:
@monopole @natasha41575 I am not sure if I agree with this behavior. In this [test case](https://github.com/kubernetes-sigs/kustomize/pull/3159/files#diff-d439562bab858ebaa57ae8c621f357b89626415b2bec9fad2b1ce0443d629c7fR424) for example, why is the expected behavior to append the new section as opposed to update the protocol to `TCP`? As mentioned in the tests we should first get to consensus on what is the expected behavior for `append vs update` before moving further. 

And if merge key can be more than one, why only tuples? Why can't it be combination of 3 keys {containerPort, protocol, name}?

Reply:
> And if merge key can be more than one, why only tuples? Why can't it be combination of 3 keys {containerPort, protocol, name}?
You can, if the openAPI schema had {containerPort, protocol, name} in "x-kubernetes-list-map-keys" 
It would also update the protocol rather than append if the openAPI schema had only {containerPort} in "x-kubernetes-list-map-keys" instead of {containerPort, protocol}

Reply:
I think what @monopole said in one of our conversations is that with this code, whether or not the behavior is correct is dependent on the keys in the schema now

Reply:
> whether or not the behavior is correct is dependent on the keys in the schema now

That seems correct.

> And if merge key can be more than one, why only tuples? Why can't it be combination of 3 keys {containerPort, protocol, name}?

The API doesn't use the name as the key because the name is too often missing, this would have broken existing behavior for many people.

> what is the expected behavior for `append vs update`

Jeff said that kustomize typically appends in that case. Removal can be done with a json patch.